llama_stack/core/conversations/conversations.py

```diff
@@ -0,0 +1,290 @@
+        await self.sql_store.update(
+            table="openai_conversations", data={"items": updated_items}, where={"id": conversation_id}
+        )
+
+        logger.info(f"Created {len(created_items)} items in conversation {conversation_id}")
+
+        # Convert created items (dicts) to proper ConversationItem types
+        adapter: TypeAdapter[ConversationItem] = TypeAdapter(ConversationItem)
+        response_items: list[ConversationItem] = [adapter.validate_python(item_dict) for item_dict in created_items]
+
+        return ConversationItemList(
+            data=response_items,
+            first_id=created_items[0]["id"] if created_items else None,
+            last_id=created_items[-1]["id"] if created_items else None,
+            has_more=False,
+        )
+
+    async def retrieve(self, conversation_id: str, item_id: str) -> ConversationItem:
+        """Retrieve a conversation item."""
+        if not conversation_id:
+            raise ValueError(f"Expected a non-empty value for `conversation_id` but received {conversation_id!r}")
+        if not item_id:
+            raise ValueError(f"Expected a non-empty value for `item_id` but received {item_id!r}")
+
+        record = await self.sql_store.fetch_one(table="openai_conversations", where={"id": conversation_id})
+        items = record.get("items", []) if record else []
+
+        for item in items:
+            if isinstance(item, dict) and item.get("id") == item_id:
+                adapter: TypeAdapter[ConversationItem] = TypeAdapter(ConversationItem)
+                return adapter.validate_python(item)
+
+        raise ValueError(f"Item {item_id} not found in conversation {conversation_id}")
+
+    async def list(self, conversation_id: str, after=NOT_GIVEN, include=NOT_GIVEN, limit=NOT_GIVEN, order=NOT_GIVEN):
+        """List items in the conversation."""
+        record = await self.sql_store.fetch_one(table="openai_conversations", where={"id": conversation_id})
+        items = record.get("items", []) if record else []
+
+        if order != NOT_GIVEN and order == "asc":
+            items = items
+        else:
+            items = list(reversed(items))
+
+        actual_limit = 20
+        if limit != NOT_GIVEN and isinstance(limit, int):
+            actual_limit = limit
+
+        items = items[:actual_limit]
+
+        # Items from database are stored as dicts, convert them to ConversationItem
+        adapter: TypeAdapter[ConversationItem] = TypeAdapter(ConversationItem)
+        response_items: list[ConversationItem] = [
+            adapter.validate_python(item) if isinstance(item, dict) else item for item in items
+        ]
+
+        # Get first and last IDs safely
+        first_id = None
+        last_id = None
+        if items:
+            first_item = items[0]
+            last_item = items[-1]
+
+            first_id = first_item.get("id") if isinstance(first_item, dict) else getattr(first_item, "id", None)
```

Comment:
uh oh what is this? why are we doing this?

Reply:
Fixed this, thanks.